language/libra-tools/writeset-transaction-generator/src/writeset_builder.rs

```diff
@@ -0,0 +1,199 @@
+                &NoContextLog::new(),
+            )
+            .unwrap()
+    }
+
+    pub fn set_libra_version(&mut self, version: Version) {
+        self.exec_func(
+            libra_root_address(),
+            "LibraVersion",
+            "set",
+            vec![],
+            vec![
+                Value::transaction_argument_signer_reference(libra_root_address()),
+                Value::u64(version),
+            ],
+        )
+    }
+}
+
+/// Convert the transaction arguments into Move values.
+fn convert_txn_args(args: &[TransactionArgument]) -> Vec<Value> {
+    args.iter()
+        .map(|arg| match arg {
+            TransactionArgument::U8(i) => Value::u8(*i),
+            TransactionArgument::U64(i) => Value::u64(*i),
+            TransactionArgument::U128(i) => Value::u128(*i),
+            TransactionArgument::Address(a) => Value::address(*a),
+            TransactionArgument::Bool(b) => Value::bool(*b),
+            TransactionArgument::U8Vector(v) => Value::vector_u8(v.clone()),
+        })
+        .collect()
+}
+
+fn move_module_changes(modules: &[CompiledModule]) -> MoveChanges {
+    let mut shadow_changeset = MoveChanges::new();
+    for module in modules.iter() {
+        let id = module.self_id();
+        let mut module_bytes = vec![];
+        module.serialize(&mut module_bytes).unwrap();
+        shadow_changeset.publish_or_overwrite_module(id, module_bytes);
+    }
+    shadow_changeset
+}
+
+fn shadow_reconfiguration<S: StateView>(state_view: &S, move_changes: &mut MoveChanges) {
+    let move_vm = MoveVM::new();
+    let state_view_storage = RemoteStorage::new(state_view);
+    let mut session = move_vm.new_session(&state_view_storage);
+    session
+        .execute_function(
+            &ModuleId::new(
+                account_config::CORE_CODE_ADDRESS,
+                Identifier::new("LibraConfig").unwrap(),
+            ),
+            &Identifier::new("reconfigure_").unwrap(),
+            vec![],
+            vec![],
+            libra_root_address(),
+            &mut CostStrategy::system(&ZERO_COST_SCHEDULE, GasUnits::new(100_000_000)),
+            &NoContextLog::new(),
+        )
+        .unwrap();
+
+    let effect = session.finish().unwrap();
+    for (account, ops) in effect.resources {
+        for (struct_tag, op) in ops {
+            match op {
+                Some((layout, value)) => {
+                    let blob = value.simple_serialize(&layout).unwrap();
+                    move_changes
+                        .publish_resource(account, struct_tag, blob)
+                        .unwrap();
+                }
+                None => move_changes
+                    .unpublish_resource(account, struct_tag)
+                    .unwrap(),
+            }
+        }
+    }
+}
+
+pub fn build_changeset<S: StateView, F>(
+    state_view: &S,
+    override_move_changes: Option<MoveChanges>,
+    procedure: F,
+    to_be_published_modules: &[CompiledModule],
```

Comment:
One question is if we would like to shadow the existing modules on chain with the modules that we are planning to publish by default. I can see arguments for both points:
- Shadow because we want to be able to invoke init functions in those modules.
- Not shadow, as it will give more flexibility in case if some functions in the old modules need to be invoked.

The current api uses the none shadow version.

Reply:
I wonder why you're taking the modules to be published as an array instead of allowing them to be published by the procedure? In other words, what's your concern about adding a `fn publish_module` to `GenesisSession`? To me this seems to be both simpler and more expressive, as now you can use these modules during the process.

I get that you want to check if the new modules are compatible. However it's still possible to do that if they are published by the procedure (see my comment below).

Reply:
Great question! The reason why I deliberatly removed `publish_module` from `GenesisSession` is because `publish_module` requires that module to be not existent on chain in VM currently. If we change the semantic of that function we will need to change the way how Libra Adapter works, which is sth I don't want to touch at the moment.

Reply:
> because publish_module requires that module to be not existent on chain in VM currently
Technically this is not true. `move_vm_test_utils::ChangeSet` allows you to pretend a module is deleted through shadowing. You also have the option to generate a `move_vm_test_utils::ChangeSet` (or WriteSet) and then merge it directly with another ChangeSet that contains the module you want to overwrite.